### PR TITLE
Further harden CSRF token comparisons

### PR DESCRIPTION
**Pull request by markstory**
Nirmal Kirubakaran contacted us via the security mailing list and disclosed a potential vulnerability in our CSRF token generation. Currently our CSRF token validation will accept any matching token pairs. If an attacker were to use an XSS vulnerabiity or physical access to fixate a CSRF token they could then exploit additional CSRF attacks.

This change improves token generation to use a random and hmac component to ensure that the token was originally generated by the receiving application.

I've put this for 4.0.6 as it is fix for a security related issue. However, I'm also ok moving it to 4.1.0 as it could potentially break multi-application setups where CSRF tokens are shared across applications or CakePHP versions < 4.0.6

### CODE REVIEW

File: src/Http/Middleware/CsrfProtectionMiddleware.php [hunk 1]
```diff
@@ -223,16 +240,57 @@ protected function _validateToken(ServerRequestInterface $request): void
...
+        }
+        $postKey = (string)substr($post, 0, static::TOKEN_VALUE_LENGTH);
+        $postHmac = (string)substr($post, static::TOKEN_VALUE_LENGTH);
+        $cookieKey = (string)substr($cookie, 0, static::TOKEN_VALUE_LENGTH);
+        $cookieHmac = (string)substr($cookie, static::TOKEN_VALUE_LENGTH);
+
+        // Put all checks in a list
+        // so they all burn time reducing timing attack window.
+        $checks = [
+            hash_equals($postKey, $cookieKey),
```
Comment (othercorey) [inline src/Http/Middleware/CsrfProtectionMiddleware.php:280]: Wouldn't this always be the same if both $post and $cookie are ''?

**Comment by ADmad** [comment]
> However, I'm also ok moving it to 4.1.0 as it could potentially break multi-application setups where CSRF tokens are shared across applications or CakePHP versions < 4.0.6

I think chances of this are very small. Just in case someone does have an issue, switching to their own middleware which follows old implementation would be pretty easy.

File: src/Http/Middleware/CsrfProtectionMiddleware.php (hunk 1 above)
Comment (markstory) [reply src/Http/Middleware/CsrfProtectionMiddleware.php:280]: Yes it would. But we need to run all the same comparisons to reduce timing attack windows.

Comment (othercorey) [reply src/Http/Middleware/CsrfProtectionMiddleware.php:280]: I mean won't it be a false positive?

Comment (garas) [reply src/Http/Middleware/CsrfProtectionMiddleware.php:281]: `_validateToken()` has `if (!$cookie)` check for empty cookie.

But check for cookie/post minimal length could be done too.

Comment (othercorey) [reply src/Http/Middleware/CsrfProtectionMiddleware.php:281]: If they aren't long enough, substr can generate false which is cast to string which could be the same also.

Comment (othercorey) [reply src/Http/Middleware/CsrfProtectionMiddleware.php:281]: You would somehow need a key whose hmac matches `(string)false` which doesn't seem feasible.  Looks good to me.

**Pull request merged**